File: evm/python/examples/send_transfer.py
"""Send native currency from the vault, directly as the role's authority."""

import json
import os

from eth_account import Account
from web3 import Web3
from web3.middleware import SignAndSendRawMiddlewareBuilder

from swig_evm import get_contract

web3 = Web3(Web3.HTTPProvider(os.environ["RPC_URL"]))
signer = Account.from_key(os.environ["SIGNER_PRIVATE_KEY"])
web3.middleware_onion.inject(SignAndSendRawMiddlewareBuilder.build(signer), layer=0)
web3.eth.default_account = signer.address
config = get_contract(web3, "SwigConfig", os.environ["ACCOUNT_ADDRESS"])
# Empty authorization selects direct EOA authentication; the caller pays gas.
hash_value = config.functions.signV2(
    int(os.environ.get("ROLE_ID", "0")),
    Web3.to_checksum_address(os.environ["RECIPIENT"]),
    int(os.environ["VALUE_WEI"]),
    b"",
    b"",
).transact()
receipt = web3.eth.wait_for_transaction_receipt(hash_value)
if receipt["status"] != 1:
    raise RuntimeError("Transfer reverted")
print(json.dumps({"transactionHash": Web3.to_hex(hash_value)}))

File: evm/rust/examples/send_transfer.rs
//! Send native currency from the vault, directly as the role's authority.
use alloy::primitives::Bytes;
use alloy::{
    network::EthereumWallet,
    primitives::{Address, U256},
    providers::ProviderBuilder,
    signers::local::PrivateKeySigner,
};
use std::{env, error::Error};
use swig_evm::contracts::config::SwigConfig;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let signer: PrivateKeySigner = env::var("SIGNER_PRIVATE_KEY")?
        .parse()
        .map_err(|_| "invalid signing key")?;
    let provider = ProviderBuilder::new()
        .wallet(EthereumWallet::from(signer))
        .connect_http(env::var("RPC_URL")?.parse()?);
    let address: Address = env::var("ACCOUNT_ADDRESS")?.parse()?;
    let config = SwigConfig::new(address, provider);
    let role_id = env::var("ROLE_ID").unwrap_or_else(|_| "0".into()).parse()?;
    let recipient = env::var("RECIPIENT")?.parse()?;
    let value: U256 = env::var("VALUE_WEI")?.parse()?;
    // Empty authorization selects direct EOA authentication; the caller pays gas.
    let receipt = config
        .signV2(role_id, recipient, value, Bytes::new(), Bytes::new())
        .send()
        .await?
        .get_receipt()
        .await?;
    if !receipt.status() {
        return Err("transfer reverted".into());
    }
    println!(
        "{}",
        serde_json::json!({"transactionHash": receipt.transaction_hash})
    );
    Ok(())
}
